### ISSUE TITLE
Jupyterlab terminal does not recognize my conda environment

### ISSUE THREAD

**Issue by ericmjl** [jupyterlab/jupyterlab#5290]
In the Jupyterlab terminal, I tried activating my conda environment:

```bash
$ source activate nav
(nav)
$ which python
~/anaconda/bin/python
```

The correct Python path, which is correctly returned from a bash session in iTerm, should be:
```bash
$ which python
~/anaconda/envs/nav/bin/python
```

Is there a fix for this, and if not, is there something that a beginner to the Jupyterlab codebase like me can help with fixing this issue?

**Comment by saulshanabrook** [comment]
Since this issue is about conda and isn't JupyterLab specific, I would raise it at https://github.com/conda/conda/issues. Good luck!

**Comment by ericmjl** [comment]
@saulshanabrook I think there might be a misunderstanding here.

The source activate command works in my regular terminal, but not in the Jupyterlab terminal emulator. Is it still correct to say that this is a conda issue, and not a jupyterlab issue?

**Comment by saulshanabrook** [comment]
@ericmjl Yep, I missed that part, sorry about that. 

It might be related to https://github.com/jupyterlab/jupyterlab/issues/3094?

**Comment by ericmjl** [comment]
Perhaps so, but is this something that is fixable at the moment, or is it a thing where I should wait on it for a while? I'm happy either way, just wanted to know :smile:.

**Comment by saulshanabrook** [comment]
I have also had this happen to me in the VS Code terminal... I am not quite sure why. I would say wait until either that other issue is fixed or until someone can figure out why this is happening.

**Comment by jasongrout** [comment]
What if you use the new recommended `conda activate`?

**Comment by saulshanabrook** [comment]
That doesn't work for me.

**Comment by jasongrout** [comment]
> That doesn't work for me.

Just checking: `which conda` returns the same as `which conda` in your normal terminal?

Also, how to do you get conda in your path in the first place? I'm wondering if there are actually two different condas that are being activated here.

**Comment by Kabongosalomon** [comment]
Was this issue solved?

**Comment by sterlinm** [comment]
I'm having a similar issue with conda and jupyter lab on os x.

I have 3 conda environments on my machine:
- base (/Users/msterling/miniconda3)
- lab (/Users/msterling/miniconda3/envs/lab)
- py38 (/Users/msterling/miniconda3/envs/py38)

I have jupyter lab installed in the `lab` environment and the kernel for py38 is installed in the kernels for the lab environment (/Users/msterling/miniconda3/envs/lab/share/jupyter/kernels).

Essentially, in the desktop terminal the path in conda environments is $CONDA_PATH:$PATH while in the terminal in Jupyter Lab the path is $PATH:$CONDA_PATH so it uses the built in python at /usr/bin/python.

Also, if I launch a console or notebook from Jupyter Lab and select one of the environments then the path looks correct and it uses the appropriate python version for each environment.

From my computer's terminal the paths look correct in each environment:

```
(base) $ which python
/Users/msterling/miniconda3/bin/python
(base) $ echo $PATH
/Users/msterling/miniconda3/bin:/Users/msterling/miniconda3/condabin:/usr/local/bin:/usr/bin:/bin
(base) $ conda activate lab
(lab) $ which python
/Users/msterling/miniconda3/envs/lab/bin/python
(lab) $ echo $PATH
/Users/msterling/miniconda3/envs/lab/bin:/Users/msterling/miniconda3/condabin:/usr/local/bin:/usr/bin:/bin:/usr/sbin
(lab) $ conda activate py38
(py38) $ which python
/Users/msterling/miniconda3/envs/py38/bin/python
(py38) $ echo $PATH
/Users/msterling/miniconda3/envs/py38/bin:/Users/msterling/miniconda3/condabin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
```

When I launch jupyterlab (from the `lab` environment) and open the terminal things breakdown:

```
(base) $ which python
/usr/bin/python
(base) $ echo $PATH
/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin:/Users/msterling/miniconda3/bin:/Users/msterling/miniconda3/condabin
(base) $ conda activate py38
(py38) $ which python
/usr/bin/python
(py38) $ echo $PATH
/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/msterling/miniconda3/envs/py38/bin:/Users/msterling/miniconda3/condabin
```

I've tried to see if there's some weird bash profile thing lingering that's causing the issue but I can't find it. I'm launching jupyter from a conda environment that's not the base environment because I thought that would be safer but suppose that could be causing issues.

**Comment by c-l-nguyen** [comment]
Ok I was able to solve this on my machine by resetting the PATH variable to be the default at the top of my .zshrc file before anything else touches PATH. That is, I have include two lines at the top:

```
PATH=/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
export PATH
```

After restarting Jupyter Lab, doing `which python` now gives me the correct python version instead of 2.7 like it was before. Shoutout to @sterlinm for your comment since I figured if the PATH was flipped, I could just force it to be default to resolve that.

**Comment by sterlinm** [comment]
Explicitly setting the path in my .zhrc file fixed this for me as well, although I'd love to understand what's going on. It's possible I just missed it, but I wonder if this might be an area where if I figured out how some of this works a bit better it would be helpful for me to add some more details to the jupyter docs. I'm always afraid that the documentation already exists somewhere in the jupyter ecosystem and I'm just not finding it.

**Comment by krbnite** [comment]
I had JupyterLab/terminal problem as @sterlinm.  

In my case, the correct python interpreter was used when invoking `python` in a regular terminal shell (e.g., `$HOME/anaconda3/bin/python`), but not from terminals opened within a JupyterLab session.  I found that once inside a JupyterLab session, the PATH variable no longer had `$HOME/anaconda3/bin` in front (or anywhere at all, actually) and so the python interpreter was defaulting to `/usr/bin/python`.  

For a while, I finagled with my .bashrc file (and even trying out the .zshrc stuff above), but had no success.  One thing I noticed is that almost everything else from my .bash_profile seemed to get picked up inside the JupyterLab terminal.  I also found that typing in `bash` once inside the JupyterLab terminal fixed this issue.

Ultimately, I found that the following if-statement provided by the Anaconda installation (automatically injected into my .bash_profile when Anaconda was installed) was the culprit; though it works in a regular terminal, it does not appear to work when starting a new terminal inside a JupyterLab session:

```
# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('$HOME/anaconda3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "$HOME/anaconda3/etc/profile.d/conda.sh" ]; then
        . "$HOME/anaconda3/etc/profile.d/conda.sh"
    else
        export PATH="$HOME/anaconda3/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<
```

THE FIX:  Adding the following if-statement after Conda's default if-statement in my .bash_profile fixed things.

```
if [ `echo $PATH | cut -d: -f1 | rev | cut -d/ -f2 | rev` != anaconda3 ]; then 
  export PATH="$HOME/anaconda3/bin:$HOME/anaconda3/condabin:$PATH"
fi
```

In a regular terminal session, the Anaconda3 paths are already in front, so they don't need to be added again.  In the JupyterLab session, when the Conda's default if-statement fails to put these Anaconda3 paths into the PATH variable, then my subsequent if-statement picks up the slack.  

This worked for me.  Hopefully it's helpful for someone else down the road as well.

**Comment by stessaris** [comment]
I believe that it's a problem of the `conda` bash function, because once I directly use the `__conda_activate` the function starts to behave (jupyterlab terminal on mac):

```bash
$ conda activate xxxx
(xxxx) $ which python
/usr/local/opt/python/libexec/bin/python
(xxxx) $ conda deactivate
$ which python
/usr/local/opt/python/libexec/bin/python
$ __conda_activate activate xxxx
(xxxx) $ which python
/Users/user/.conda/envs/xxxx/bin/python
(xxxx) $ conda deactivate
$ which python
/usr/local/opt/python/libexec/bin/python
$ conda activate xxxx
(xxxx) $ which python
/Users/user/.conda/envs/xxxx/bin/python
(xxxx) $ 
```

I think it might be related to a variable that is not properly initialised

**Comment by wcneill** [comment]
I do not know if this is directly or only tangentially related, but my Jupyter Lab terminal can't see `conda` at all. Attempting to activate an environment from within Jupyter Lab terminal gets me a `bash: activate: No such file or directory`. 

I installed `nb_conda` and can now choose to run a kernel in a specific environment but I have no access to the `conda` command line utility from within Jupyter Lab. 

Should I start a new issue, or is this the correct one to file this behavior under?

**Comment by GrigoriiTarasov** [comment]
> I had JupyterLab/terminal problem as @sterlinm.
> 
> In my case, the correct python interpreter was used when invoking `python` in a regular terminal shell (e.g., `$HOME/anaconda3/bin/python`), but not from terminals opened within a JupyterLab session. I found that once inside a JupyterLab session, the PATH variable no longer had `$HOME/anaconda3/bin` in front (or anywhere at all, actually) and so the python interpreter was defaulting to `/usr/bin/python`.
> 
> For a while, I finagled with my .bashrc file (and even trying out the .zshrc stuff above), but had no success. One thing I noticed is that almost everything else from my .bash_profile seemed to get picked up inside the JupyterLab terminal. I also found that typing in `bash` once inside the JupyterLab terminal fixed this issue.
> 
> Ultimately, I found that the following if-statement provided by the Anaconda installation (automatically injected into my .bash_profile when Anaconda was installed) was the culprit; though it works in a regular terminal, it does not appear to work when starting a new terminal inside a JupyterLab session:
> 
> ```
> # >>> conda initialize >>>
> # !! Contents within this block are managed by 'conda init' !!
> __conda_setup="$('$HOME/anaconda3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
> if [ $? -eq 0 ]; then
>     eval "$__conda_setup"
> else
>     if [ -f "$HOME/anaconda3/etc/profile.d/conda.sh" ]; then
>         . "$HOME/anaconda3/etc/profile.d/conda.sh"
>     else
>         export PATH="$HOME/anaconda3/bin:$PATH"
>     fi
> fi
> unset __conda_setup
> # <<< conda initialize <<<
> ```
> 
> THE FIX: Adding the following if-statement after Conda's default if-statement in my .bash_profile fixed things.
> 
> ```
> if [ `echo $PATH | cut -d: -f1 | rev | cut -d/ -f2 | rev` != anaconda3 ]; then 
>   export PATH="$HOME/anaconda3/bin:$HOME/anaconda3/condabin:$PATH"
> fi
> ```
> 
> In a regular terminal session, the Anaconda3 paths are already in front, so they don't need to be added again. In the JupyterLab session, when the Conda's default if-statement fails to put these Anaconda3 paths into the PATH variable, then my subsequent if-statement picks up the slack.
> 
> This worked for me. Hopefully it's helpful for someone else down the road as well.

Hello, let me ask some usage conditions.  My ~/.bash_profile was empy,
adding only 

```
if [ `echo $PATH | cut -d: -f1 | rev | cut -d/ -f2 | rev` != anaconda3 ]; then 
  export PATH="$HOME/anaconda3/bin:$HOME/anaconda3/condabin:$PATH"
fi
```
haven't worked.

Adding it with the whole block
```
# >>> conda initialize >>>
...
priorly also. Do I need some other lines in ~/.bash_profile or to delete smth in ~/.bashrc
